sql/src/main/java/io/crate/execution/engine/window/WindowFunctionBatchIterator.java

```diff
@@ -183,8 +195,32 @@ public boolean hasNext() {
                     idxInPartition = 0;
                     pEnd = findFirstNonPeer(sortedRows, pStart, end, cmpPartitionBy);
                 }
-                int wBegin = pStart; // UNBOUNDED PRECEDING -> Frame always starts at the start of the partition
-                int wEnd = findFirstNonPeer(sortedRows, i, pEnd, cmpOrderBy);
+
+                int wBegin;
+                int wEnd;
+
+                if (isUnboundedPrecedingCurrentRow) {
```

Comment:
I wonder if we could encapsulate these logic within the frameDefinition type?
So that we have something like

```
wBegin = frameDefinition.start().type().findStartBoundary(sortedRows, ...)
wEnd = frameDefinition.end().type()....
```

Seems natural to me that the type would encapsulate how the boundary is computed.
We could then probably also unit test better the logic.

Reply:
I thought about something similar but it didn't feel right to move the know-how of partitioning and ordering to a frame type, or even to a frame definition (and this logic also depends on the position in the result set `i` which makes it even trickier to isolate)

Reply:
Not sure if I understand this correctly - but I think it isn't required to move the know-how of *how* the partitioning works? Only what the current partition bounds are. 

As it is there is a lot of runtime (per row) conditionals that would be resolvable once per query execution. We can either use some form of polymorphism (e.g. by having methods on the frame definition / type) or introduce some mapping from type to higher order functions and use those.

I also think that the `start` / `end` handling should be separated. `isUnboundedPrecedingCurrentRow` is handling a concrete pair. Once we support more frame bound definitions this approach won't hold as we'll have quite a lot of possible combinations.

Reply:
Pushed a sketch on this. And fixups for the previous suggestions